nixos/modules/services/networking/tetrd.nix

```diff
@@ -0,0 +1,92 @@
+{ config, lib, pkgs, ... }:
+
+let
+  tetrd-init = pkgs.writeShellScript "tetrd-init" ''
+    mkdir -p /usr/bin
+    ln -sf ${pkgs.nettools}/bin/{route,ifconfig} /usr/bin
+  '';
+in {
+  options.services.tetrd.enable = lib.mkEnableOption pkgs.tetrd.meta.description;
+
+  config = lib.mkIf config.services.tetrd.enable {
+    environment.systemPackages = [ pkgs.tetrd ];
+
+    systemd.services.tetrd = {
+      description = pkgs.tetrd.meta.description;
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStartPre = "+${tetrd-init}";
+        ExecStart = "${pkgs.tetrd}/opt/Tetrd/bin/tetrd";
+        Restart = "always";
+        RuntimeDirectory = "tetrd";
+        RootDirectory = "/run/tetrd";
+        DynamicUser = true;
+        DeviceAllow = "usb_device";
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        NoNewPrivileges = true;
+        PrivateMounts = true;
+        PrivateNetwork = lib.mkDefault false;
+        PrivateTmp = true;
+        PrivateUsers = lib.mkDefault false;
+        ProtectClock = lib.mkDefault false;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        ProtectSystem = "strict";
+        RemoveIPC = true;
+        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" "AF_NETLINK" ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        SystemCallArchitectures = "native";
+
+        SystemCallFilter = [
+          "@system-service"
+          "~@aio"
+          "~@chown"
+          "~@clock"
+          "~@cpu-emulation"
+          "~@debug"
+          "~@keyring"
+          "~@memlock"
+          "~@module"
+          "~@mount"
+          "~@obsolete"
+          "~@pkey"
+          "~@raw-io"
+          "~@reboot"
+          "~@swap"
+          "~@sync"
+        ];
+
+        BindReadOnlyPaths = [
+          builtins.storeDir
+          "/etc/ssl"
+          "/etc/static/ssl"
+        ];
+
+        BindPaths = [
+          "/etc/resolv.conf"
```

Comment:
Binding `/etc/resolv.conf` may not work well, especially in use cases where this file is replaced by a DHCP client. Also, why not in  `BindReadOnlyPaths=`? Does `tetrd` need to modify this file? I think you should just put the whole /etc in either `BindPaths=` or `BindReadOnlyPaths=`, and write an AppArmor profile if you really need hardening at the file level.

Reply:
Yes, tetrd really does need write acces to /etc/resolv.conf.
Ironically, it doesn't seem to actually do anything to it, but it crashes if it can't write.

Reply:
Never mind, it does change.
From
```
# Generated by resolvconf
options edns0
```
To
```
nameserver 2001:4860:4860::8888
nameserver 2620:0:ccc::2
nameserver 8.8.8.8
nameserver 208.67.222.222
```

Reply:
Only hacky thing I can think of for a ``/etc/resolv.conf`` generated by nix is start: [unlink, touch], end: [rm relink] by the service.

Reply:
If the host needs to use the `resolv.conf` generated by `tetrd`, maybe you could let the user or your module do: 
```nix
environment.etc."resolv.conf".source = "/run/tetrd/etc/resolv.conf"
```

Reply:
Hmm, that would also require `/run/tetrd/etc/resolv.conf` to not be a bind of `/etc/resolv.conf`, maybe this can be done by removing and replacing `/etc/resolv.conf` in the chroot, in `preStart=`

Reply:
hmm, nah, `/run/tetrd/etc/resolv.conf` is likely not visible from the host because of `PrivateMount=`

Reply:
Maybe touching `/etc/tetrd/resolv.conf` with `systemd.tmpfiles`, and then `BindPaths="/etc/tetrd/resolv.conf:/etc/resolv.conf"` and `environment.etc."resolv.conf".source = "/etc/tetrd/etc/resolv.conf"`.
Maybe also there is no use of having a `resolv.conf` from `tetrd`, and thus `BindPaths="/dev/null:/etc/resolv.conf"` would be enough.

Reply:
Indeed, those IPv6 and IPv4 addresses are all public resolvers', so I guess it's useless to let `tetrd` messing around the real `resolv.conf`.

Reply:
``/dev/null`` didn't work unfortunately, but ``/etc/tetrd/resolv.conf`` method did work.
I'll update it in a sec.